cachi2/core/package_managers/rubygems/parser.py

```diff
@@ -0,0 +1,82 @@
+import logging
+from typing import Optional, Union
+
+import pydantic
+
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+GEMFILE_LOCK = "Gemfile.lock"
+
+
+class GemMetadata(pydantic.BaseModel, extra="forbid"):
+    """
+    Base class for gem metadata.
+
+    Attributes:
+        name:       The name of the gem.
+        version:    The version of the gem.
+    """
+
+    name: str
+    version: str
+
+
+class GemDependency(GemMetadata):
+    """
+    Represents a gem dependency.
+
+    Attributes:
+        source:     The source URL of the gem.
+        checksum:   The checksum of the gem.
+    """
+
+    source: str
+    checksum: Optional[str] = None
+
+
+class GitDependency(GemMetadata):
+    """
+    Represents a git dependency.
+
+    Attributes:
+        url:        The URL of the git repository.
+        revision:   Commit hash.
+        branch:     The branch of the git repository.
+    """
+
+    url: str
+    revision: str
+    branch: Optional[str] = None
+
+
+class PathDependency(GemMetadata):
+    """
+    Represents a path dependency.
+
+    Attributes:
+        path:       Subpath from package root.
+    """
+
+    path: str
+
+
+BundlerDependency = Union[GemDependency, GitDependency, PathDependency]
+
+
+def parse_gemlock(package_dir: RootedPath) -> list[BundlerDependency]:
+    """TODO."""
+    return NotImplemented
+
+
+def _validate_gem_dependency(gem: GemDependency) -> None:
```

Comment:
Added the docs.

Reply:
One of the [threads](https://github.com/containerbuildsystem/cachi2/pull/565#discussion_r1700321119) in the design PR pointed out that there's no real reason for us to avoid downloading deps from alternate registries. So I guess we won't be needing this check.

Reply:
Ah, sorry, I completely forgot about that.